Feed page placeholder

jimit200 opened the pull request:
This page will serve as a placeholder for the feed that a user will see. The view buttons still need to be synced to the assignment sample in the section page

AndrewLester committed:
Add initial feed tab and FeedItem model

jimit200 committed:
commented out the video player for now so the app runs

rahulshamkuwar commented on lib/src/models/feedItem.dart:
```diff
@@ -0,0 +1,34 @@
+      return FeedItemType.Homework;
+    } else if (item is Meeting) {
+      return FeedItemType.Meeting;
+    }
```
```suggestion
    } else {
      return FeedItemType.Lecture
    }
```


Just a guess from me, haven't tested to see if it would work.

AndrewLester replied:
I was looking to add this but I think I'll wait until a LectureItem is created, otherwise there won't be any way to receive such a type.

MattGeimer replied:
It's better to put this in now even if it won't function than not and hope you remember later. I can't even start to count the number of times I've done the latter to my own detriment.

rahulshamkuwar commented on lib/src/screens/home/feedTab.dart:
```diff
@@ -12,8 +21,111 @@ class FeedTab extends StatefulWidget {
+      width: size.width,
+      height: size.height,
+      child: ListView.builder(
+        padding: EdgeInsets.all(20),
```
Seems quite excessive for padding... wouldn't it be better to have something like 5 or 10?

```suggestion
```

Edit:
Actually, better to just delete it and change the padding below.

rahulshamkuwar commented on lib/src/screens/home/feedTab.dart:
```diff
@@ -12,8 +21,111 @@ class FeedTab extends StatefulWidget {
+    //}
+
+    return Padding(
+      padding: EdgeInsets.all(20),
```
```suggestion
      padding: EdgeInsets.all(5),
```

rahulshamkuwar commented on lib/src/screens/home/feedTab.dart:
```diff
@@ -12,8 +21,111 @@ class FeedTab extends StatefulWidget {
+      attachmentUrl:
+          'https://images.all-free-download.com/footage_preview/mp4/kitty_68.mp4',
+    ),
+    FeedItem(
+      name: 'Homework 5',
+      owner: 'English 101',
+      body: 'Never do this',
+      type: FeedItemType.Homework,
+      itemId: '54t3t3',
+      postDate: DateTime.now().add(Duration(hours: 2)),
+    ),
```
Might be better to use a `HomeworkItem` to display details for homework, that way the `onTap` could just pass the `HomeworkItem` to the `Homework.dart` page.

Just for future reference, right now since its hard coded, it doesn't matter.

MattGeimer replied:
Agreed. Maybe do that now but hardcode the `HomeworkItem`?

AndrewLester replied:
I chose to go with this "FeedItem" structure so that all actions that occur for a section can be added to a user's feed collection, and the specific "item" that's tied to the action doesn't have to be loaded in the feed page until it's clicked.

rahulshamkuwar replied:
Yeah, I get what you mean, but in the long run, we will need to make a get request to get the `FeedItem`, and then make another get request for the `HomeworkItem` when the user taps on it, in which case we might as well use a `HomeworkItem`.

AndrewLester replied:
Good point, I think we can discuss more in the meeting but I imagine quite a long feed. We could probably only show the first 10 or so items and get others later, but that would mean a lot of data to view the page.

rahulshamkuwar commented on lib/src/screens/home/feedTab.dart:
```diff
@@ -12,8 +21,111 @@ class FeedTab extends StatefulWidget {
+    return Padding(
+      padding: EdgeInsets.all(20),
+      child: Card(
+        child: InkWell(
```
```suggestion
        color: MediaQuery.of(context).platformBrightness == Brightness.dark
            ? Colors.grey[900]
            : Colors.grey[300],
        child: InkWell(
```

Change colors based on the device theme.

rahulshamkuwar commented on lib/src/screens/home/feedTab.dart:
```diff
@@ -12,8 +21,111 @@ class FeedTab extends StatefulWidget {
+                  item.body,
+                  textAlign: TextAlign.center,
+                ),
+                SizedBox(height: 10),
+                SizedBox(height: 10),
```
```suggestion
                SizedBox(height: 20),
```
Not sure why there are 2....

AndrewLester replied:
I think I was planning on adding something in between there, but I forgot. It may have been the video player/image container.

AndrewLester replied:
Merged

rahulshamkuwar commented on lib/src/screens/home/feedTab.dart:
```diff
@@ -12,8 +21,111 @@ class FeedTab extends StatefulWidget {
+                          ],
+                        ),
+                        color: Colors.white,
+                        cupertino: (context, platform) => CupertinoButtonData(
+                            color: AppColors.secondaryUofILight),
+                        materialFlat: (context, platform) =>
+                            MaterialFlatButtonData(
+                                textColor: Colors.white,
+                                color: AppColors.secondaryUofILight)),
```
```suggestion
                        cupertino: (context, platform) => CupertinoButtonData(
                        color: MediaQuery.of(context).platformBrightness ==
                                Brightness.dark
                            ? AppColors.secondaryUofIDark
                            : AppColors.secondaryUofILight,
                      ),
                      materialFlat: (context, platform) =>
                          MaterialFlatButtonData(
                        textColor: Colors.white,
                        color: MediaQuery.of(context).platformBrightness ==
                                Brightness.dark
                            ? AppColors.secondaryUofIDark
                            : AppColors.secondaryUofILight,
                      ),
```
Change colors based on the device theme.

MattGeimer replied:
This looks quite long and messy. I think it'd be good if the `PlatformButton` was refactored out so it was only one line.

rahulshamkuwar reviewed (changes requested):
Looks good, just a few changes are needed.

MattGeimer commented on lib/src/models/feedItem.dart:
```diff
@@ -0,0 +1,34 @@
+  }
+}
+
+class FeedItem {
+  String name;
+  String owner;
+  String body;
+  DateTime postDate;
+  FeedItemType type;
+  String itemId;
+  String? attachmentUrl;
```
Add docs. For example, how do I know what an `itemId` is?

MattGeimer commented on lib/src/screens/home/feedTab.dart:
```diff
@@ -12,8 +21,111 @@ class FeedTab extends StatefulWidget {
+
+  Widget buildItem(BuildContext context, FeedItem item) {
+    VideoPlayerController? videoPlayerController;
+    // if (item.type == FeedItemType.Lecture && item.attachmentUrl != null) {
+    //   videoPlayerController =
+    //       VideoPlayerController.network(item.attachmentUrl!);
+    //   await videoPlayerController.initialize();
+    //}
```
Remove dead code

AndrewLester committed:
Add feed page navigation by making other screens loadable

AndrewLester committed:
Create Button widget for PlatformButton

AndrewLester committed:
Remove dead code

AndrewLester committed:
Fix padding again

AndrewLester committed:
Change button color based on screen brightness

MattGeimer commented on lib/src/screens/courses/specificCourseView.dart:
```diff
@@ -254,18 +254,18 @@ class _CourseViewState extends State<CourseView> {
-                  ),
-                );
-              },
+              // onTap: () {
+              //   Navigator.of(context, rootNavigator: true).push(
+              //     CupertinoPageRoute(
+              //       builder: (context) {
+              //         return SectionView(
+              //           sectionName: courseItem,
+              //           sectionId: courseItem.sections[index].sectionID,
+              //         );
+              //       },
+              //     ),
+              //   );
+              // },
```
```suggestion
```

Remove dead code.

MattGeimer commented on lib/src/screens/courses/specificSectionView.dart:
```diff
@@ -1,89 +1,100 @@
+  _SectionViewState() {
+    // widget.course.then((course) {
+    //   homework = [
+    //     HomeworkItem(
+    //       dueDate: DateTime.now().add(Duration(days: 2)),
+    //       name: 'Practice Problems #1',
+    //       description: 'This homework will help prepare you for the test!',
+    //       assignmentUrl: 'https://example.com',
+    //       platform: 'turnitin',
+    //       course: course,
+    //       files: [
+    //         File(
+    //           name: 'Problem set.json',
+    //           mimeType: 'application/json',
+    //           size: 400,
+    //           url: 'https://example.com',
+    //         ),
+    //         File(
+    //           name: 'Problem set.json',
+    //           mimeType: 'application/json',
+    //           size: 400,
+    //           url: 'https://example.com',
+    //         ),
+    //         File(
+    //           name: 'Problem set.json',
+    //           mimeType: 'application/json',
+    //           size: 400,
+    //           url: 'https://example.com',
+    //         )
+    //       ],
+    //     ),
+    //     HomeworkItem(
+    //       name: 'Practice Problems #2',
+    //       dueDate: DateTime.now().add(Duration(days: 7)),
+    //       assignmentUrl: 'https://en.wikipedia.org/wiki/Hot_air_ballooning',
+    //       platform: 'turnitin',
+    //       course: course,
+    //     )
+    //   ];
+    // });
```
```suggestion
```

Remove dead code.

MattGeimer commented on lib/src/screens/courses/specificSectionView.dart:
```diff
@@ -184,20 +195,47 @@ class _SectionViewState extends State<SectionView> {
+          } else {
+            final SectionItem section = data[0];
+            final CourseItem course = data[1];
+            // return _getDetails(section);
```
```suggestion
```

Remove dead code.

MattGeimer commented on lib/src/screens/home/feedTab.dart:
```diff
@@ -12,8 +24,140 @@ class FeedTab extends StatefulWidget {
+                          Icon(PlatformIcons(context).eyeSolid),
+                          SizedBox(width: 5),
+                          PlatformText('View'),
+                        ],
+                      ),
+                    ),
+                  ],
+                )
+              ],
+            ),
+          ),
+        ),
+      ),
+    );
+  }
```
This is a sign that some of this should ideally be refactored out into smaller, simpler views.

MattGeimer reviewed (commented):
Lots of dead code that needs to be cleaned up. Otherwise LGTM!

rahulshamkuwar commented on lib/src/screens/home/search.dart:
```diff
@@ -75,7 +75,7 @@ class Search extends SearchDelegate<CourseItem> {
 
   _getCourses() {
-    return _courseAPI.getCourses(query, false);
+    return _courseAPI.getCourses(query);
```
Just wondering, why did you remove the bool?

rahulshamkuwar replied:
Ah nvm, I see what you did.

AndrewLester committed:
Remove instances of dead code and refactor feedtab build method

rahulshamkuwar commented on lib/src/screens/home/feedTab.dart:
```diff
@@ -12,8 +19,144 @@ class FeedTab extends StatefulWidget {
+            ? Colors.grey[900]
+            : Colors.grey[300],
+        child: InkWell(
+          onTap: () => Navigator.of(context).push(
+            platformPageRoute(
+              context: context,
+              builder: (context) {
+                return _getScreenForFeedItem(item);
+              },
+            ),
+          ),
```
```suggestion
          onTap: () => Navigator.of(context, rootNavigator: true).push(
            CupertinoPageRoute(
              builder: (context) {
                return _getScreenForFeedItem(item);
              },
            ),
          ),
```
I know I've been pushing for using multiplatform stuff, but for page routes, I think it's better to use CupertinoPageRoute, they feel more intuitive to use than MaterialPageRoute from a user's perspective. Let me know what you think.

rahulshamkuwar commented on lib/src/screens/home/feedTab.dart:
```diff
@@ -12,8 +19,144 @@ class FeedTab extends StatefulWidget {
+          Row(
+            children: [
+              InkWell(
+                onTap: () => Navigator.of(context).push(
+                  platformPageRoute(
+                    context: context,
+                    builder: (context) {
+                      return SectionView(
+                        sectionName: item.owner,
+                        sectionCode: item.itemCode,
+                      );
+                    },
+                  ),
```
```suggestion
                onTap: () => Navigator.of(context, rootNavigator: true).push(
                  CupertinoPageRoute(
                    builder: (context) {
                      return SectionView(
                        sectionName: item.owner,
                        sectionCode: item.itemCode,
                      );
                    },
                  ),
```

rahulshamkuwar commented on lib/src/screens/home/feedTab.dart:
```diff
@@ -12,8 +19,144 @@ class FeedTab extends StatefulWidget {
+                ),
+              Spacer(),
+              Button(
+                onPressed: () => Navigator.of(context).push(
+                  platformPageRoute(
+                    context: context,
+                    builder: (context) {
+                      return _getScreenForFeedItem(item);
+                    },
+                  ),
+                ),
```
```suggestion
                onPressed: () => Navigator.of(context, rootNavigator: true).push(
                  CupertinoPageRoute(
                    builder: (context) {
                      return _getScreenForFeedItem(item);
                    },
                  ),
                ),
```

rahulshamkuwar commented on lib/src/screens/homework/homeworkScreen.dart:
```diff
@@ -1,28 +1,87 @@
+            if (snapshot.connectionState == ConnectionState.waiting ||
+                data == null) {
+              return Center(
+                child: CircularProgressIndicator(
+                  color: AppColors.urbanaOrange,
```
Take a look at the `SpinKitRing` that I used in `search.dart`

rahulshamkuwar commented on lib/src/screens/courses/specificSectionView.dart:
```diff
@@ -184,20 +195,46 @@ class _SectionViewState extends State<SectionView> {
+          if (snapshot.connectionState == ConnectionState.waiting ||
+              data == null) {
+            return Center(
+              child: CircularProgressIndicator(
+                color: AppColors.urbanaOrange,
+              ),
```
Take a look at the `SpinKitRing` that I used in `search.dart`

rahulshamkuwar commented on lib/src/widgets/button.dart:
```diff
@@ -0,0 +1,31 @@
+    return PlatformButton(
+      onPressed: onPressed,
+      child: child,
+      color: Colors.white,
```
Probably make this a property like the `child` property, just in case. Set the default to white.

rahulshamkuwar commented on lib/src/providers/courseApi.dart:
```diff
@@ -1,25 +1,47 @@
     return courses;
   }
+
+  Future<SectionItem> getSection(String fullCode) async {
+    List<String> codeSections = fullCode.split('_');
+    String code = codeSections[0] + codeSections[1];
+    int crn = int.parse(codeSections[2]);
+    Uri uri = Uri.http(Config.baseEndpoint!, '/api/v1/section/search',
+        {'code': code, 'CRN': crn.toString()});
+
+    final response = await client.get(uri);
+    if (response.statusCode != 200) {
+      throw HttpException(
+          response.reasonPhrase ?? response.statusCode.toString());
+    }
+
+    dynamic data = jsonDecode(response.body)['sections'][0];
+    SectionItem section = SectionItem.fromJSON(data);
+    return section;
+  }
```
Going to talk about this in the vc.

AndrewLester committed:
Use CupertinoPageRoute instead of MaterialPageRoute

Co-authored-by: Rahul Shamkuwar <45727404+rahulshamkuwar@users.noreply.github.com>

AndrewLester committed:
Use CupertinoPageRoute instead of MaterialPageRoute for button

Co-authored-by: Rahul Shamkuwar <45727404+rahulshamkuwar@users.noreply.github.com>

AndrewLester committed:
Use CupertinoPageRoute instead of MaterialPageRoute for section click

Co-authored-by: Rahul Shamkuwar <45727404+rahulshamkuwar@users.noreply.github.com>

AndrewLester committed:
Use SpinKitRing and provide parameter for Button color

AndrewLester committed:
Increase post date contrast for Feed items

AndrewLester committed:
Replace constructor with initState

AndrewLester committed:
Initial update to the new feed API

AndrewLester committed:
Use InheritedWidget for different app servvices

AndrewLester committed:
Call ApiService::login to run API login endpoint when firebase auth is succesfull

AndrewLester committed:
Remove API provider

AndrewLester committed:
Fix widgets to work with new API layout

AndrewLester committed:
Use sub navigator and change model schemaas

AndrewLester committed:
Fix type error using List.from

AndrewLester committed:
Use proper setState call and fix code parsing

rahulshamkuwar commented on lib/src/models/courseItem.dart:
```diff
@@ -99,7 +99,7 @@ class CourseItem implements Comparable {
       classScheduleInformation: json["classScheduleInformation"] ?? null,
       sections: _getSections(json["sections"]),
-      categories: [json["genEd"]],
+      categories: List<String>.from(["genEd"]),
```
```suggestion
      categories: List<String>.from(json["genEd"]),
```
Correct me if I'm wrong, but I think it would be like this?

rahulshamkuwar commented on lib/src/onePlace.dart:
```diff
@@ -58,7 +61,17 @@ class _OnePlaceState extends State<OnePlace> {
           value: FirebaseAuthService().userStream,
           initialData: FirebaseAuthService().user,
-          child: SplashScreen(),
+          child: Services(
+            child: Navigator(
+              initialRoute: 'splashScreen',
+              onGenerateRoute: (settings) {
+                if (settings.name == 'splashScreen') {
+                  return CupertinoPageRoute(
+                      builder: (context) => SplashScreen());
+                }
+              },
+            ),
+          ),
```
Haven't looked into the `Services` class, what does it do?

rahulshamkuwar commented on lib/src/models/feedItem.dart:
```diff
@@ -0,0 +1,65 @@
+  String action;
+
+  /// The code of the item referenced by this feed item
+  String itemCode;
```
@jwparktom @AndrewLester Perhaps we could have the API return a corresponding `SectionItem` instead of an `itemCode`, that way we don't need to create another request when opening the section from this `FeedItem`?

AndrewLester replied:
I can cover this a bit more tomorrow, but using the SectionItem returned in the feed list requests (as it currently returns sections as well) was unfortunately not enough, since the section view also requires a course.

We could return sections AND courses from the feed endpoint. It may be better to discuss this further tomorrow though and create a solution that works elsewhere too.

rahulshamkuwar replied:
Oh yeah, I forgot about that.... hmm... If I'm not mistaken, the only use for the `CourseItem` was to show the title, perhaps we can get that from a `SectionItem` itself by changing the model?

AndrewLester replied:
That's a fair idea, my only issue with that is we may not need it in other cases, so it would appear to be an odd piece of data to leave in the SectionItem. It may already be in the section item though to be honest. In fact, I think it *is*, inside the section's `full_code`. We can talk about this later, maybe we should write some helper methods to turn full codes -> subject ID/course ID

rahulshamkuwar commented on lib/src/screens/courses/specificSectionView.dart:
```diff
@@ -184,20 +206,46 @@ class _SectionViewState extends State<SectionView> {
+            );
+          } else if (snapshot.hasError) {
+            final error = snapshot.error;
+            return Center(
+              child: Text(error.toString()),
+            );
```
```suggestion
            return Center(
              child: AlertBox(
                child: Text(
                  error.toString(),
                  style: TextStyle(color: Colors.white),
                ),
              ),
            );
```
Perhaps use the `AlertBox` class for any errors?

rahulshamkuwar commented on lib/src/screens/homework/homeworkScreen.dart:
```diff
@@ -1,28 +1,81 @@
+                ),
+              );
+            } else if (snapshot.hasError) {
+              return Center(
+                child: Text(snapshot.error.toString()),
+              );
```
```suggestion
              return Center(
              child: AlertBox(
                child: Text(
                  snapshot.error.toString(),
                  style: TextStyle(color: Colors.white),
                ),
              ),
            );
```
Perhaps use the `AlertBox` class for any errors?

rahulshamkuwar reviewed (changes requested):
Really good work here!! Just have the main comment to clarify, about sending a `SectionItem` with the `FeedItem` so that there are fewer API requests being made. Let me know what you think about that

AndrewLester commented on lib/src/screens/courses/specificSectionView.dart:
```diff
@@ -1,89 +1,109 @@
+
+    course.then((course) async {
+      var homework = await homeworkApi.getHomework('code');
+
```
```suggestion
if (!mounted) return;
```
Make sure the widget is still mounted when setting state after this await.

AndrewLester committed:
Move Services to top level

AndrewLester committed:
Fix specific section view to use new course system and add alert boxes

rahulshamkuwar committed:
Created Dio client.
Uses Provider to make a globally avaiable Dio client. It should be used for all requests that needs the cookie, since the cookie will already be attached. Anything that doesn't need the cookie can just create a new local Dio client instead.

rahulshamkuwar committed:
Merge branch 'feed-test' into api-test

rahulshamkuwar committed:
Removed inherited widgets.
Also removed boilerplate code.
Implemented the use of AccountProvider, but havent been able to test it yet due to API changes.

rahulshamkuwar committed:
bumb version

rahulshamkuwar committed:
Updated AccountProvider.
Added comments for more models.
Minor UI changes.

rahulshamkuwar committed:
Merge pull request #35 from SiGMobileUIUC/api-test

Created Dio client.
